File: python/refrase/registry.py
"""Config-driven model registry.

Loads all built-in family configs at import time and builds a lookup index.
Supports runtime registration of custom models and families.
"""

import json
from importlib import resources
from typing import Any, Optional


# ── Internal state ──

_registry: dict[str, dict[str, Any]] = {}  # model_id -> RegistryEntry
_families: dict[str, dict[str, Any]] = {}  # family_name -> FamilyConfig


def _load_configs() -> dict[str, dict[str, Any]]:
    """Load all built-in JSON family configs from the configs/ package directory."""
    configs: dict[str, dict[str, Any]] = {}
    config_dir = resources.files("refrase") / "configs"
    for path in config_dir.iterdir():
        if hasattr(path, "name") and path.name.endswith(".json"):
            config = json.loads(path.read_text())
            configs[config["family"]] = config
    return configs


def _register_family_internal(config: dict[str, Any]) -> None:
    """Register a family and all its models in the internal registry."""
    _families[config["family"]] = config
    for model_id, model_cfg in config["models"].items():
        entry = {
            "family_config": config,
            "model_config": model_cfg,
            "model_id": model_id,
        }
        _registry[model_id] = entry
        # Register aliases
        for alias in model_cfg.get("aliases") or []:
            _registry[alias] = entry


def _init() -> None:
    """Initialize registry from built-in configs."""
    for config in _load_configs().values():
        _register_family_internal(config)


# Initialize on load
_init()


# ── Internal API ──

def _get_entry(model_id: str) -> tuple[dict[str, Any], dict[str, Any], str]:
    """Look up a registry entry by model ID.

    Returns:
        Tuple of (family_config, model_config, canonical_model_id).

    Raises:
        ValueError: If model_id is not found.
    """
    entry = _registry.get(model_id)
    if entry is None:
        available = ", ".join(sorted(_registry.keys()))
        raise ValueError(
            f'Unknown model: "{model_id}". Available models: {available}'
        )
    return entry["family_config"], entry["model_config"], entry["model_id"]


# ── Public API ──

def list_models() -> list[dict[str, Any]]:
    """List all registered models, sorted by ID.

    Returns:
        List of dicts with id, family, variant, name, provider.
    """
    seen: set[str] = set()
    entries: list[dict[str, Any]] = []

    for id_, entry in _registry.items():
        # Skip aliases (where the registered ID differs from the canonical modelId)
        if id_ != entry["model_id"]:
            continue
        if id_ in seen:
            continue
        seen.add(id_)

        entries.append({
            "id": id_,
            "family": entry["family_config"]["family"],
            "variant": entry["model_config"]["variant"],
            "name": entry["model_config"].get("name"),
            "provider": entry["family_config"].get("provider"),
        })

    return sorted(entries, key=lambda e: e["id"])


def list_families() -> list[dict[str, Any]]:
    """List all registered families with summary info.

    Returns:
        List of dicts with family, provider, docs_url, model_count, rule_count.
    """
    infos: list[dict[str, Any]] = []
    for config in _families.values():
        infos.append({
            "family": config["family"],
            "provider": config["provider"],
            "docs_url": config.get("docs_url"),
            "model_count": len(config["models"]),
            "rule_count": len(config["rules"]),
        })
    return sorted(infos, key=lambda i: i["family"])


def get_model_config(model_id: str) -> dict[str, Any]:
    """Get the full config for a specific model.

    Returns:
        Dict with model config fields plus family, provider, docs_url.
    """
    family_config, model_config, _ = _get_entry(model_id)
    return {
        **model_config,
        "family": family_config["family"],
        "provider": family_config["provider"],
        "docs_url": family_config.get("docs_url"),
    }


def get_family_config(family: str) -> dict[str, Any]:
    """Get the full family config.

    Raises:
        ValueError: If family is not found.
    """
    import copy
    config = _families.get(family)
    if config is None:
        available = ", ".join(sorted(_families.keys()))
        raise ValueError(
            f'Unknown family: "{family}". Available families: {available}'
        )
    return copy.deepcopy(config)


def register_model(
    family: str,
    model_id: str,
    config: dict[str, Any],
) -> None:
    """Register a custom model at runtime, associating it with an existing family.

    The model will inherit the family's rules and API hints.

    Args:
        family: The family name to associate the model with.
        model_id: The model identifier.
        config: Dict with name, variant, and optionally tier and aliases.

    Raises:
        ValueError: If the family is not found.
    """
    family_config = _families.get(family)
    if family_config is None:
        raise ValueError(
            f'Cannot register model "{model_id}": unknown family "{family}". '
            f"Register the family first with register_family()."
        )

    model_config = {
        "name": config["name"],
        "variant": config["variant"],
        "tier": config.get("tier"),
        "aliases": config.get("aliases"),
    }

    # Add to the family config's models
    family_config["models"][model_id] = model_config

    # Add to registry
    entry = {
        "family_config": family_config,
        "model_config": model_config,
        "model_id": model_id,
    }
    _registry[model_id] = entry

    for alias in config.get("aliases") or []:
        _registry[alias] = entry


def register_family(config: dict[str, Any]) -> None:
    """Register a custom model family at runtime with its own rules.

    If the family already exists, it will be overwritten.
    """
    _register_family_internal(config)

File: typescript/src/registry.ts
/**
 * Config-driven model registry.
 *
 * Loads all built-in family configs at import time and builds a lookup index.
 * Supports runtime registration of custom models and families.
 */
import { CONFIGS } from "./configs/index.js";
import type {
  FamilyConfig,
  FamilyInfo,
  ModelConfig,
  ModelEntry,
  ModelRegistration,
} from "./types.js";

interface RegistryEntry {
  familyConfig: FamilyConfig;
  modelConfig: ModelConfig;
  modelId: string;
}

const _registry: Map<string, RegistryEntry> = new Map();
const _families: Map<string, FamilyConfig> = new Map();

/** Initialize registry from built-in configs. */
function _init(): void {
  for (const config of Object.values(CONFIGS)) {
    _registerFamilyInternal(config);
  }
}

function _registerFamilyInternal(config: FamilyConfig): void {
  _families.set(config.family, config);
  for (const [modelId, modelCfg] of Object.entries(config.models)) {
    const entry: RegistryEntry = {
      familyConfig: config,
      modelConfig: modelCfg,
      modelId,
    };
    _registry.set(modelId, entry);
    // Register aliases
    for (const alias of modelCfg.aliases ?? []) {
      _registry.set(alias, entry);
    }
  }
}

// Initialize on load
_init();

// ── Internal API ──

/** Look up a registry entry by model ID. Throws if not found. */
export function _getEntry(modelId: string): RegistryEntry {
  const entry = _registry.get(modelId);
  if (!entry) {
    const available = [..._registry.keys()].sort().join(", ");
    throw new Error(
      `Unknown model: "${modelId}". Available models: ${available}`,
    );
  }
  return entry;
}

// ── Public API ──

/** List all registered models, sorted by ID. */
export function listModels(): ModelEntry[] {
  // Deduplicate aliases — only include canonical model IDs
  const seen = new Set<string>();
  const entries: ModelEntry[] = [];

  for (const [id, entry] of _registry) {
    // Skip aliases (where the registered ID differs from the canonical modelId)
    if (id !== entry.modelId) continue;
    if (seen.has(id)) continue;
    seen.add(id);

    entries.push({
      id,
      family: entry.familyConfig.family,
      variant: entry.modelConfig.variant,
      name: entry.modelConfig.name,
      provider: entry.familyConfig.provider,
    });
  }

  return entries.sort((a, b) => a.id.localeCompare(b.id));
}

/** List all registered families with summary info. */
export function listFamilies(): FamilyInfo[] {
  const infos: FamilyInfo[] = [];
  for (const config of _families.values()) {
    infos.push({
      family: config.family,
      provider: config.provider,
      docsUrl: config.docs_url,
      modelCount: Object.keys(config.models).length,
      ruleCount: config.rules.length,
    });
  }
  return infos.sort((a, b) => a.family.localeCompare(b.family));
}

/** Get the full config for a specific model. */
export function getModelConfig(
  modelId: string,
): ModelConfig & { family: string; provider: string; docsUrl?: string } {
  const entry = _getEntry(modelId);
  return {
    ...entry.modelConfig,
    family: entry.familyConfig.family,
    provider: entry.familyConfig.provider,
    docsUrl: entry.familyConfig.docs_url,
  };
}

/** Get the full family config (returns a deep copy to prevent accidental mutation). */
export function getFamilyConfig(family: string): FamilyConfig {
  const config = _families.get(family);
  if (!config) {
    const available = [..._families.keys()].sort().join(", ");
    throw new Error(
      `Unknown family: "${family}". Available families: ${available}`,
    );
  }
  return JSON.parse(JSON.stringify(config)) as FamilyConfig;
}

/**
 * Register a custom model at runtime, associating it with an existing family.
 * The model will inherit the family's rules and API hints.
 */
export function registerModel(
  familyName: string,
  modelId: string,
  config: ModelRegistration,
): void {
  const familyConfig = _families.get(familyName);
  if (!familyConfig) {
    throw new Error(
      `Cannot register model "${modelId}": unknown family "${familyName}". ` +
        `Register the family first with registerFamily().`,
    );
  }

  const modelConfig: ModelConfig = {
    name: config.name,
    variant: config.variant,
    tier: config.tier,
    aliases: config.aliases,
  };

  // Add to the family config's models
  familyConfig.models[modelId] = modelConfig;

  // Add to registry
  const entry: RegistryEntry = {
    familyConfig,
    modelConfig,
    modelId,
  };
  _registry.set(modelId, entry);

  for (const alias of config.aliases ?? []) {
    _registry.set(alias, entry);
  }
}

/**
 * Register a custom model family at runtime with its own rules.
 * If the family already exists, it will be overwritten.
 */
export function registerFamily(config: FamilyConfig): void {
  _registerFamilyInternal(config);
}
